### PR TITLE
Support multiple destination folders per App instance

### PR DESCRIPTION
**Pull request by Averus-a**
### Description of Change ###
Contains refactoring/changes to ensure the behavior is still correct if you define multiple destination folders for App (like separate app/libs folders)

### Issues Resolved ### 

- fixes #

### API Changes ###
 
None

### Platforms Affected ### 

- All

### Behavioral Changes ###

None

### PR Checklist ###

- [ ] I have included examples or tests
- [ ] I have updated the change log
- [ ] I am listed in the CONTRIBUTORS file
- [ ] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: src/Orc.NuGetExplorer.Example/ProjectManagement/ExampleProjects/ExampleProject.cs [hunk 1]
```diff
@@ -29,6 +29,10 @@ public ExampleProject(IFrameworkNameProvider frameworkNameProvider)
 
         public ImmutableList<NuGetFramework> SupportedPlatforms { get; set; }
 
+        public bool CanIgnoreDependencies { get; } = true;
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer.Example/ProjectManagement/ExampleProjects/ExampleProject.cs:32]: Maybe we should remove the Can, just name it "IgnoreDependencies" and "SupportSideBySide"

File: src/Orc.NuGetExplorer.Xaml/ViewModels/PackageDetailsViewModel.cs [hunk 2]
```diff
@@ -312,7 +300,7 @@ private async Task ApplyPackageAsync()
 
                 SelectedVersion = selectedVersion;
 
-                PackageMetadataProvider = InitMetadataProvider();
+                PackageMetadataProvider = Providers.PackageMetadataProvider.CreateFromSourceContext(ServiceLocator.Default);
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer.Xaml/ViewModels/PackageDetailsViewModel.cs:303]: ServiceLocator.Default or should we inject?

File: src/Orc.NuGetExplorer/Management/ExtensibleProjectLocator.cs [hunk 3]
```diff
@@ -82,14 +82,19 @@ public void Register(IExtensibleProject project)
 
         public void Register<T>() where T : IExtensibleProject
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Management/ExtensibleProjectLocator.cs:83]: where T should go on new line

File: src/Orc.NuGetExplorer/Management/ExtensibleProjectLocator.cs [hunk 4]
```diff
@@ -82,14 +82,19 @@ public void Register(IExtensibleProject project)
...
         {
-            Register<T>(new object[] { });
+            Register(_typeFactory.CreateInstance<T>());
         }
 
         public void Register<T>(params object[] parameters) where T : IExtensibleProject
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Management/ExtensibleProjectLocator.cs:88]: put where : T on new line

File: src/Orc.NuGetExplorer/Management/Projects/Interfaces/IExtensibleProject.cs [hunk 5]
```diff
@@ -17,12 +17,10 @@ public interface IExtensibleProject
 
         PackagePathResolver GetPathResolver();
 
-        string GetInstallPath(PackageIdentity packageIdentity);
-
-        void Install();
+        bool CanIgnoreDependencies { get; }
 
-        void Update();
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Management/Projects/Interfaces/IExtensibleProject.cs:24]: This is a breaking change but I think it does not impact us

File: src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs [hunk 6]
```diff
@@ -21,47 +22,68 @@ public class PackageMetadataProvider : IPackageMetadataProvider
...
 
         static PackageMetadataProvider()
         {
             NuGetLogger = ServiceLocator.Default.ResolveType<ILogger>();
         }
 
+        public PackageMetadataProvider(ISourceRepositoryProvider repositoryProvider)
+        {
+            Argument.IsNotNull(() => repositoryProvider);
+            _repositoryProvider = repositoryProvider;
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs:42]: empty line between argument check and field assignment

File: src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs [hunk 7]
```diff
@@ -21,47 +22,68 @@ public class PackageMetadataProvider : IPackageMetadataProvider
...
             Argument.IsNotNull(() => directoryService);
             Argument.IsNotNull(() => sourceRepositories);
 
             _directoryService = directoryService;
             _sourceRepositories = sourceRepositories;
             _optionalLocalRepositories = optionalGlobalLocalRepositories;
-            _localRepository = localRepository;
         }
 
+        public static PackageMetadataProvider CreateFromSourceContext(IServiceLocator locator)
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs:67]: rename to serviceLocator

File: src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs [hunk 8]
```diff
@@ -21,47 +22,68 @@ public class PackageMetadataProvider : IPackageMetadataProvider
...
+
+            return PackageMetadataProvider.CreateFromSourceContext(directoryService, repositoryService, projectSource, packageManager);
+        }
 
-        public static PackageMetadataProvider CreateFromSourceContext(IDirectoryService directoryService, IRepositoryContextService repositoryService, IExtensibleProjectLocator projectSource, INuGetPackageManager projectManager)
+        public static PackageMetadataProvider CreateFromSourceContext(IDirectoryService directoryService, IRepositoryContextService repositoryService, IExtensibleProjectLocator projectSource, 
+            INuGetPackageManager projectManager)
         {
             Argument.IsNotNull(() => directoryService);
             Argument.IsNotNull(() => repositoryService);
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs:83]: should we null check project manager, etc?

File: src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs [hunk 9]
```diff
@@ -83,6 +105,25 @@ public async Task<IPackageSearchMetadata> GetLocalPackageMetadataAsync(PackageId
                 sources.AddRange(_optionalLocalRepositories);
             }
 
+            if (_localRepository is null)
+            {
+                // Support multiple destinations
+                if (_localRepository is null)
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Providers/PackageMetadataProvider.cs:111]: same check twice? I think this should check for something else

File: src/Orc.NuGetExplorer/Services/PackageInstallationService.cs [hunk 10]
```diff
@@ -136,6 +141,12 @@ internal class PackageInstallationService : IPackageInstallationService
             {
                 foreach (var removedPackage in uninstalledPackages)
                 {
+                    if (removedPackage.Version is null)
+                    {
+                        _nugetLogger.LogWarning($"Skip package {removedPackage.Id} uninstall. Check your package.config for references on this packages");
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Services/PackageInstallationService.cs:146]: "Check your package.config for references of this packages"

File: src/Orc.NuGetExplorer/Services/PackageOperationService.cs [hunk 11]
```diff
@@ -126,27 +126,32 @@ public async Task UpdatePackagesAsync(IPackageDetails package, bool allowedPrere
             //create current version identity
             var currentVersion = await _nuGetPackageManager.GetVersionInstalledAsync(_defaultProject, updateIdentity.Id, token);
             var currentIdentity = new PackageIdentity(updateIdentity.Id, currentVersion);
+            if (currentIdentity.Version is null)
+            {
+                // Can be because of mismatch between packages.config and package files
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Services/PackageOperationService.cs:131]: should we log here?

**Review by GeertvanHorrik — CHANGES_REQUESTED** [review]
Please see comments.

Also try to add some more unit tests if possible.

File: src/Orc.NuGetExplorer.Xaml/ViewModels/PackageDetailsViewModel.cs (hunk 2 above)
Comment (Averus-a) [reply src/Orc.NuGetExplorer.Xaml/ViewModels/PackageDetailsViewModel.cs:303]: I have a refactorings ready in separate branch which are get rid of this weird provider initialization, so decided to just put ServiceLocator here for now. This part should be cleared off in the near future

File: src/Orc.NuGetExplorer/Services/PackageOperationService.cs [hunk 12]
```diff
@@ -123,30 +123,38 @@ public async Task UpdatePackagesAsync(IPackageDetails package, bool allowedPrere
...
 
-            //create current version identity
+            // create current version identity
             var currentVersion = await _nuGetPackageManager.GetVersionInstalledAsync(_defaultProject, updateIdentity.Id, token);
             var currentIdentity = new PackageIdentity(updateIdentity.Id, currentVersion);
+            if (currentIdentity.Version is null)
+            {
+                // Can be because of mismatch between packages.config and package files
+
+                _logger.LogWarning($"Cannot found existing local files for installed '{package.Id}'. Continue update to version '{updateIdentity.Version}'");
```
Comment (GeertvanHorrik) [inline src/Orc.NuGetExplorer/Services/PackageOperationService.cs:133]: Could not find existing local files for installed ...

**Review by GeertvanHorrik — APPROVED** [review]

**Pull request closed without merging**